compiler/main/log.cpp

```diff
@@ -45,37 +49,74 @@ bool             fdump_html_print_block_IDs             =   false;
 FILE*            deletedIdHandle                        =    NULL;
 char             deletedIdFilename[FILENAME_MAX + 1]    =      "";
 
-static Vec<char> valid_log_flags;
-static Vec<char> log_flags;
+// Keeping names of available passes
+static bool availableInitialized = false;
+std::vector<std::string> logAvailableNames;
+std::vector<char> logAvailableShortNames;
 
-void log_flags_arg(const ArgumentDescription* desc, const char* arg) {
-  if (valid_log_flags.count() == 0)
-    initLogFlags(valid_log_flags);
+// Which passes should be logged?
+static bool logAll = true;
+std::set<std::string> logOnlyName;
 
-  // --log or --log= means "log every pass"
-  // Actually, passes whose log letter is NUL are skipped.
-  if (*arg == '\0') {   // empty
-    log_flags.set_union(valid_log_flags);   // Set all.
 
-  } else {
+void logMakePassAvailable(const char* name, char shortname)
+{
+  logAvailableNames.push_back(name);
+  logAvailableShortNames.push_back(shortname);
+}
 
-    // Parse the argument of --log=<arg> or -d<arg> for specific log flags.
-    // Each flag is a single letter; see log.h for the correspondence.
-    while (*arg) {
-      if (valid_log_flags.set_in(*arg))
-        log_flags.set_add(*arg);
-      else {
-        fprintf(stderr, "Unrecognized log flag: '%c'\n", *arg);
-        clean_exit(1);
-      }
+void logSelectPass(const char* arg) {
+  // The first time this is run, collect the available pass names.
+  if (availableInitialized == false) {
+    initPassesForLogging(); // calls logMakePassAvailable for each pass
+    availableInitialized = true;
+  }
+
+  // Make --log optional by setting it if --log-pass= is set
+  fLog = true;
+
+  // Since we selected a pass, don't log everything
+  logAll = false;
+
+  // Check first for a match in the available pass names
+  for( size_t i = 0; i < logAvailableNames.size(); i++ ) {
+    if (logAvailableNames[i] == arg) {
+      logOnlyName.insert(logAvailableNames[i]);
+      return;
+    }
+  }
 
-      arg++;
+  // If none found, check for a match in the short names.
+  if (strlen(arg) == 1 &&
+      arg[0] != LOG_NO_SHORT &&
+      arg[0] != LOG_NEVER) {
+    for( size_t i = 0; i < logAvailableShortNames.size(); i++ ) {
+      if (logAvailableShortNames[i] == arg[0]) {
+        logOnlyName.insert(logAvailableNames[i]);
```

Comment:
It no longer supports e.g. --log-pass=RS. Instead you can pass --log-pass=R --log-pass=S. In that event, this routine is called twice, once with "R" and once with "S". The routine first checks for pass names matching. Since none will be 1-character long, it does this second check.